### PR TITLE
elastic-agent diagnostics pprof

### PR DESCRIPTION
**Pull request by michel-laterman**
## What does this PR do?

Add a new `Pprof` grpc call as well as two ways to gather pprof data.
1. Collect all pprof data from the agent and the beats it runs with
```
elastic-agent diagnostics collect --pprof
```

2. Collect one pprof profile from one application-(route) with
```
elastic-agent diagnostics pprof [--pprof-application APP-NAME] [--pprof-route-key ROUTE-KEY] [--pprof-type TYPE]
```

The duration of the `trace` or `profile` data can be controlled with the `pprof-duration` flag, set to `30s` by default. The timeout for these commands can be controlled separately with the `timeout` flag, by default this value is `30s + pprof-duration`. `timeout` should be larger then `pprof-duration` to allow the command to process the response data.

## Why is it important?

This option allows operators to gather pprof data from all applications into the archive in order to assist debugging.
Adds the `elastic-agent diagnostics pprof` command to allow operators to gather one pprof type for debugging (specifically to be composed with other tools on the command line).

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation [PR here](https://github.com/elastic/observability-docs/pull/1284)
- [] ~~I have made corresponding change to the default configuration files~~
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## How to test this PR locally

Run `elastic-agent collect --pprof` and inspect the archive for pprof data.

## Related issues

- Relates #21965

### CODE REVIEW

**Comment by michel-laterman** [comment]
@ruflin here's a draft of adding pprof support.

I think we should consider adding a config option in order to serve the pprof endpoints on the standard port (I'm not sure why it's purely done via a command line flag at the moment, is there historical reasons?)

Also let me know if I should split this PR up as it's gotten pretty large

**Comment by ruflin** [comment]
++ on also providing a config option. I always appreciate smaller PR's but leave it up to you if the overhead is worth it.

File: x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go
Comment (ruflin) [inline x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go:113]: Is my understanding correct that by default this command will take 30s to execute? Maybe we should mention this in the docs here that it is tied to pprof-duration?

File: x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go
Comment (ruflin) [inline x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go:172]: Why is it 2* ?

File: x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go
Comment (ruflin) [inline x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go:262]: Should a potential error here to be handled?

File: x-pack/elastic-agent/pkg/agent/control/server/server.go
Comment (ruflin) [inline x-pack/elastic-agent/pkg/agent/control/server/server.go:217]: @michalpristas Seems some quite fundamental changes here. Can you take a look?

File: x-pack/elastic-agent/pkg/agent/control/server/server.go
Comment (ruflin) [inline x-pack/elastic-agent/pkg/agent/control/server/server.go:247]: One more place where we have `*2`. Curious to understand this better.

File: x-pack/elastic-agent/pkg/agent/control/server/server.go
Comment (ruflin) [reply x-pack/elastic-agent/pkg/agent/control/server/server.go:217]: I see further down, most of the code is equal but split up in multiple methods which is a nice cleanup.

**Review by ruflin — COMMENTED** [review]
Overall LGTM, would be good if @michalpristas could take a closer look at the code.

One thing that concerns me but is not only related to this PR is that there are no tests added and no tests changed, meaning we don't have any. I know testing the pprof part itself might be tricky but I would hope at least some parts around creating the correct paths, building endpoints etc. could be tested.

File: x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go
Comment (michel-laterman) [reply x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go:172]: i needed a context timeout that would be larger then the trace duration to give the agent time to process/zip the responses. I can change this to a static addition (i.e. 30s + pprofDur)

File: x-pack/elastic-agent/pkg/agent/control/server/server.go
Comment (michel-laterman) [reply x-pack/elastic-agent/pkg/agent/control/server/server.go:247]: I think that the duration passed to `newSocketRequester` is an artifact of my development (I was not doing requests with contexts at one point); i'll clean this up

**Comment by michel-laterman** [comment]
@ruflin, I think the diagnostics commands are a good candidate for adding an e2e test. I'm not sure if we should validate archive structure, i think just ensuring the command exited with a 0 and that an archive was created may be goo enough for now.

File: x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go
Comment (ruflin) [reply x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go]: Got it. Can you add this comment to the code so someone else touching this code will know why this is there? What approach you take (30s or 2x) I leave to you.

**Comment by ruflin** [comment]
👍 on adding an e2e for the diagnostic command. Can you follow up on this in the e2e repo after this got in?

Should this be marked as ready for review? There are some CI complaints ..

File: x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go
Comment (michel-laterman) [reply x-pack/elastic-agent/pkg/agent/cmd/diagnostics.go]: I'll use the 30s+duration value, but I've added an option to specify the top level command timeout so it will be an option to increase it beyond this limit if that's needed. Hopefully with this we can avoid edge cases where it takes the agent over 30s to create an archive.

File: x-pack/elastic-agent/CHANGELOG.next.asciidoc
Comment (ruflin) [inline x-pack/elastic-agent/CHANGELOG.next.asciidoc:153]: Lets make sure we follow up with proper docs.

File: x-pack/elastic-agent/pkg/agent/control/server/server.go
Comment (ruflin) [inline x-pack/elastic-agent/pkg/agent/control/server/server.go:266]: In the past we tried to use Warn as little as possible as it is not clear if the user needs to take action or not. What are the reasons we can't retrieve the route key?

File: x-pack/elastic-agent/pkg/agent/control/server/server.go
Comment (ruflin) [inline x-pack/elastic-agent/pkg/agent/control/server/server.go:285]: We have here 3 for loops nested inside each other. Maybe we could extract some into methods which would make reviewing and potentially testing easier.

**Review by ruflin — APPROVED** [review]
LGTM

@ferullo Does endpoint have something similar to pprof which could also be read out in the future?

**Comment by ruflin** [comment]
I did try to run this locally by running `./elastic-agent run`. Likely because of this I get the following errors for pprof:

```
Get "http://unix/debug/pprof/profile?seconds=30": dial unix /tmp/elastic-agent/93926c22716b6d271825bfa3d3c9f0b3434619940c4a3ccba25d96ff2322946f.sock: connect: no such file or directory
```

Is it expected to also work if I have not installed elastic-agent but only run it?

After I run `./elastic-agent diagnostics collect --pprof` for 30s nothing happens which of course is expected. But before we start the diagnostics run, could we print out a message to std that now collection starts and will likely take x seconds?

File: x-pack/elastic-agent/CHANGELOG.next.asciidoc
Comment (michel-laterman) [reply x-pack/elastic-agent/CHANGELOG.next.asciidoc:153]: already in a pr

File: x-pack/elastic-agent/pkg/agent/control/server/server.go
Comment (michel-laterman) [reply x-pack/elastic-agent/pkg/agent/control/server/server.go:266]: I'm not sure why we would be unable to retrieve routes in this case, but I thought logging any issue could help with debugging/edge cases

File: x-pack/elastic-agent/pkg/agent/control/server/server.go
Comment (michel-laterman) [reply x-pack/elastic-agent/pkg/agent/control/server/server.go:285]: i'll work on it

File: x-pack/elastic-agent/pkg/agent/control/server/server.go
Comment (ruflin) [reply x-pack/elastic-agent/pkg/agent/control/server/server.go]: 👍

**Comment by michel-laterman** [comment]
Right now there's a bug with how the agent handles the `agent.monitoring.pprof` attribute. The output of `NewMonitor` https://github.com/elastic/beats/blob/27d1a6e3e7b1ae6eb75ed4a2f699c9a02bdab88d/x-pack/elastic-agent/pkg/core/monitoring/beats/beats_monitor.go#L35
references the attribute in the file. However, the non-nill config passed to reload 
 https://github.com/elastic/beats/blob/27d1a6e3e7b1ae6eb75ed4a2f699c9a02bdab88d/x-pack/elastic-agent/pkg/core/monitoring/beats/beats_monitor.go#L49 has the (default) `false` setting. I'm trying to find out why this is happening.

**Comment by michel-laterman** [comment]
I think the issue with the pprof attribute is that on startup in fleet mode, the agent gets a policy update from fleet in https://github.com/elastic/beats/blob/191a0752b5ceddc7b7657a517b90ca76c1350f30/x-pack/elastic-agent/pkg/agent/application/pipeline/actions/handlers/handler_action_policy_change.go#L75
The policy sent by fleet does not have a reference to the pprof attribute

File: x-pack/elastic-agent/pkg/core/monitoring/config/config.go
Comment (michel-laterman) [inline x-pack/elastic-agent/pkg/core/monitoring/config/config.go:42]: @ruflin, I changed this from a default false behaviour to default nil so that a value inserted into fleet.yml would be used when the `Reload` function is called (so the setting is now applied to beats that the agent starts)

Comment (ruflin) [reply x-pack/elastic-agent/pkg/core/monitoring/config/config.go:42]: Interesting, so value only gets overwritten if not set yet?

@narph This likely should also feed into your thoughts on how these files should work in the future.

**Pull request merged**
